File: www/FlyyPlugin.js
    var exec = require('cordova/exec');

    exports.initSdk = function (partner_token, enviroment, success, error) {
        exec(success, error, 'FlyyPlugin', 'initSDK', [partner_token, enviroment]);
    };

    exports.startOfferActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'openOfferActivity', [segmentId]);
    };

    exports.startRewardsActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'openRewardsActivity', [segmentId]);
    };

    exports.startWalletActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'openWalletActivity', [segmentId]);
    };

    exports.startGiftCardsActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'openGiftCardsActivity', [segmentId]);
    };

    exports.setUser = function (ext_uid,includeReferralCode,segmentId, success, error) {
        exec(success, error, 'FlyyPlugin', 'setUser', [ext_uid,includeReferralCode,segmentId]);
    };

    exports.setNewUser = function (ext_uid, success, error) {
        exec(success, error, 'FlyyPlugin', 'setNewUser', [ext_uid]);
    };

    exports.setUsername = function (user_name, success, error) {
        exec(success, error, 'FlyyPlugin', 'setUsername', [user_name]);
    };

    exports.trackEvent = function (key, value, success, error) {
        exec(success, error, 'FlyyPlugin', 'trackEvent', [key, value]);
    };

    exports.trackEventJson = function (key, jsonObj, success, error) {
        exec(success, error, 'FlyyPlugin', 'trackEvent', [key, JSON.stringify(jsonObj)]);
    };
    exports.setContactNumber = function (key,success, error) {
        exec(success, error, 'FlyyPlugin', 'setContactNumber', [key]);
    };
    exports.startStampActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'startStampActivity', [segmentId]);
    };
    exports.startReferralHistoryActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'startReferralHistoryActivity', [segmentId]);
    };
    exports.startTournamentListActivity = function (title,segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'startTournamentListActivity', [title,segmentId]);
    };
    exports.setRewardGridSpanCount = function (gridSpanCount,success, error) {
        exec(success, error, 'FlyyPlugin', 'setRewardGridSpanCount', [gridSpanCount]);
    };
    exports.openDeeplink = function (action,success, error) {
        exec(success, error, 'FlyyPlugin', 'openDeeplink', [action]);
    };
    exports.setAppPackage = function (packageName,success, error) {
        exec(success, error, 'FlyyPlugin', 'setAppPackage', [packageName]);
    };
    exports.handleNotification = function (remoteMessage,success, error) {
        exec(success, error, 'FlyyPlugin', 'handleNotification', [remoteMessage]);
    };
    exports.navigateToInviteAndEarnActivity = function (offerId, success, error) {
        exec(success, error, 'FlyyPlugin', 'navigateToInviteAndEarnActivity', [offerId]);
    };
    exports.navigateToCustomInviteAndEanActivity = function (offerId, toolbarItemsColor, success, error) {
        exec(success, error, 'FlyyPlugin', 'navigateToCustomInviteAndEanActivity', [offerId, toolbarItemsColor]);
    };
    exports.navigateToBonanzaActivity = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'navigateToBonanzaActivity', []);
    };
    exports.navigateToChallengeDetailActivity = function (offerId, success, error) {
        exec(success, error, 'FlyyPlugin', 'navigateToChallengeDetailActivity', [offerId]);
    };
    exports.setThemeColor = function (colorPrimary, colorPrimaryDark, success, error) {
        exec(success, error, 'FlyyPlugin', 'setThemeColor', [colorPrimary, colorPrimaryDark]);
    };
    exports.getShareData = function (offerId, success, error) {
        exec(success, error, 'FlyyPlugin', 'getShareData', [offerId]);
    };
    exports.getReferralCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getReferralCount', []);
    };
    exports.getScratchCardCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getScratchCardCount', []);
    };
    exports.getPreviousLeaderboardWinners = function (tag, success, error) {
        exec(success, error, 'FlyyPlugin', 'getPreviousLeaderboardWinners', [tag]);
    };
    exports.getLeaderboardParticipants = function (tag, success, error) {
        exec(success, error, 'FlyyPlugin', 'getLeaderboardParticipants', [tag]);
    };
    exports.getWalletBalance = function (walletLabel, success, error) {
        exec(success, error, 'FlyyPlugin', 'getWalletBalance', [walletLabel]);
    };
    exports.getReferrerDetails = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getReferrerDetails', []);
    };
    exports.getOffersCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getOffersCount', []);
    };
    exports.trackUIEvents = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'trackUIEvents', []);
    };
    exports.onFlyySDKClosed = function (success) {
        exec(success, 'FlyyPlugin', 'onSDKClosedWithScreenName', []);
    }
    exports.handleForeGroundNotification = function (remoteMessage,success, error) {
        exec(success, error, 'FlyyPlugin', 'handleForeGroundNotification', [remoteMessage]);
    };
    exports.handleBackGroundNotification = function (remoteMessage,success, error) {
        exec(success, error, 'FlyyPlugin', 'handleBackGroundNotification', [remoteMessage]);
    };
    exports.sendFCMTokenToServer = function (remoteMessage,success, error) {
        exec(success, error, 'FlyyPlugin', 'sendFCMTokenToServer', [remoteMessage]);
    };
    exports.handleDeeplink = function (deeplink,success, error) {
        exec(success, error, 'FlyyPlugin', 'handleDeeplink', [deeplink]);
    };
    exports.updateUsername = function (user_name,success, error) {
        exec(success, error, 'FlyyPlugin', 'updateUsername', [user_name]);
    };
    exports.loginUser = function (ext_uid, user_name,success, error) {
        exec(success, error, 'FlyyPlugin', 'loginUser', [ext_uid, user_name]);
    };
    exports.loginUserWithCallback = function (ext_uid, user_name,success, error) {
        exec(success, error, 'FlyyPlugin', 'loginUserWithCallback', [ext_uid, user_name]);
    };
    exports.initWithReferrerReceivedCallback = function (partner_id, enviroment,success, error) {
        exec(success, error, 'FlyyPlugin', 'initWithReferrerReceivedCallback', [partner_id, enviroment]);
    };
    exports.getTotalSCCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getTotalSCCount', []);
    };
    exports.getScratchedSCCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getScratchedSCCount', []);
    };
    exports.getUnscratchedSCCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getUnscratchedSCCount', []);
    };
    exports.getLockedSCCount = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'getLockedSCCount', []);
    };
    exports.setBaseFlyyDomain = function (domain,success, error) {
        exec(success, error, 'FlyyPlugin', 'setBaseFlyyDomain', [domain]);
    };
    exports.setBaseStageFlyyDomain = function (domain,success, error) {
        exec(success, error, 'FlyyPlugin', 'setBaseStageFlyyDomain', [domain]);
    };
    exports.startReferAndEarnActivity = function (success, error) {
        exec(success, error, 'FlyyPlugin', 'startReferAndEarnActivity', []);
    };
    exports.startCheckInActivity = function (segmentId,success, error) {
        exec(success, error, 'FlyyPlugin', 'startCheckInActivity', [segmentId]);
    };
    exports.startRetailerIncentiveApp = function (partnerId,userToken,themeColor,url,success, error) {
        exec(success, error, 'FlyyPlugin', 'startRetailerIncentiveApp', [partnerId,userToken,themeColor,url]);
    };
    exports.setCustomFontName = function (regularFont,mediumFont,boldFont,success, error) {
        exec(success, error, 'FlyyPlugin', 'setCustomFontName', [regularFont,mediumFont,boldFont]);
    };
    exports.setBankDetails = function (number,ifsc,name,success, error) {
        exec(success, error, 'FlyyPlugin', 'setBankDetails', [number,ifsc,name]);
    };
    exports.setUpi = function (upiId,success, error) {
        exec(success, error, 'FlyyPlugin', 'setUpi', [upiId]);
    };

File: src/ios/FlyyPlugin.swift
import FlyyFramework

@objc (FlyyPlugin) class FlyyPlugin: CDVPlugin {
    
    @objc(intSdk:)
    func initSdk(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let partnerToken = (command.arguments[0] as? NSObject)?.value(forKey: "partner_token") as? String
        let environment = (command.arguments[1] as? NSObject)?.value(forKey: "environment") as? String
        
        if partnerToken != nil && !partnerToken!.isEmpty && environment != nil && !environment!.isEmpty {
            if environment == "stage" {
                Flyy.sharedInstance.initSDK(partnerToken: partnerToken!, environment: Flyy.FLYY_ENVIRONMENT_STAGING)
            } else if environment == "production" {
                Flyy.sharedInstance.initSDK(partnerToken: partnerToken!, environment: Flyy.FLYY_ENVIRONMENT_PRODUCTION)
            }
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected enviroment variable as either 'stage' or 'production'."
            )
        }
    }
    
    @objc(openOfferActivity:)
    func openOfferActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let segmentId = (command.arguments[0] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if segmentId != nil && !segmentId!.isEmpty {
            Flyy.sharedInstance.setSegmentId(segementId: segmentId!)
            Flyy.sharedInstance.openOffersPage(navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
        
    }
    
    @objc(openRewardsActivity:)
    func openRewardsActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let segmentId = (command.arguments[0] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if segmentId != nil && !segmentId!.isEmpty {
            Flyy.sharedInstance.setSegmentId(segementId: segmentId!)
            Flyy.sharedInstance.openRewardsPage(navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
        
    }
    
    @objc(openWalletActivity:)
    func openWalletActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let segmentId = (command.arguments[0] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if segmentId != nil && !segmentId!.isEmpty {
            Flyy.sharedInstance.setSegmentId(segementId: segmentId!)
            Flyy.sharedInstance.openWalletPage(navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(openGiftCardsActivity:)
    func openGiftCardsActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let segmentId = (command.arguments[0] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if segmentId != nil && !segmentId!.isEmpty {
            Flyy.sharedInstance.setSegmentId(segementId: segmentId!)
            Flyy.sharedInstance.openGiftCardsPage(navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(setUser:)
    func setUser(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let extUid = (command.arguments[0] as? NSObject)?.value(forKey: "ext_uid") as? String
        let segmentId = (command.arguments[1] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if extUid != nil && !extUid!.isEmpty {
            Flyy.sharedInstance.setUser(externalUserId: extUid!, segmentId: segmentId!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(setNewUser:)
    func setNewUser(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let extUid = (command.arguments[0] as? NSObject)?.value(forKey: "ext_uid") as? String
        
        if extUid != nil && !extUid!.isEmpty {
            Flyy.sharedInstance.setNewUser(externalUserId: extUid!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(setUsername:)
    func setUsername(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let userName = (command.arguments[0] as? NSObject)?.value(forKey: "user_name") as? String
        
        if userName != nil && !userName!.isEmpty {
            Flyy.sharedInstance.setUserName(userName: userName!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(trackEvent:)
    func trackEvent(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let key = (command.arguments[0] as? NSObject)?.value(forKey: "key") as? String
        let value = (command.arguments[1] as? NSObject)?.value(forKey: "value") as? String
        
        if key != nil && !key!.isEmpty && value != nil && !value!.isEmpty {
            Flyy.sharedInstance.sendEvent(key: key!, value: value!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected two non-empty string argument."
            )
        }
    }
    
    @objc(sendNotificationReceived:)
    func sendNotificationReceived(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(notificationClicked:)
    func notificationClicked(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(setContactNumber:)
    func setContactNumber(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(startStampActivity:)
    func startStampActivity(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(startReferralHistoryActivity:)
    func startReferralHistoryActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let segmentId = (command.arguments[0] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if segmentId != nil && !segmentId!.isEmpty {
            Flyy.sharedInstance.setSegmentId(segementId: segmentId!)
            Flyy.sharedInstance.openReferralHistoryPage(navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(startTournamentListActivity:)
    func startTournamentListActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let title = (command.arguments[0] as? NSObject)?.value(forKey: "title") as? String
        let segmentId = (command.arguments[1] as? NSObject)?.value(forKey: "segmentId") as? String
        
        if title != nil && !title!.isEmpty && segmentId != nil && !segmentId!.isEmpty {
            Flyy.sharedInstance.setSegmentId(segementId: segmentId!)
            Flyy.sharedInstance.openTournamentsListPage(navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected two non-empty string argument."
            )
        }
    }
    
    @objc(setRewardGridSpanCount:)
    func setRewardGridSpanCount(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(openDeeplink:)
    func openDeeplink(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(setAppPackage:)
    func setAppPackage(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let packageName = (command.arguments[0] as? NSObject)?.value(forKey: "package_name") as? String
        
        if packageName != nil && !packageName!.isEmpty {
            Flyy.sharedInstance.setPackage(packageName: packageName!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(handleNotification:)
    func handleNotification(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let remoteMessage = (command.arguments[0] as? NSObject)?.value(forKey: "remote_message") as? String
        
        if remoteMessage != nil && !remoteMessage!.isEmpty {
            Flyy.sharedInstance.handleNotification(userInfo: remoteMessage!, navigationController: self.viewController.navigationController!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
        
    }
    
    @objc(navigateToInviteAndEarnActivity:)
    func navigateToInviteAndEarnActivity(command: CDVInvokedUrlCommand) {
        
    }
    
    @objc(navigateToCustomInviteAndEanActivity:)
    func navigateToCustomInviteAndEanActivity(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let offerId = (command.arguments[0] as? NSObject)?.value(forKey: "offer_id") as? String
        let toolbasItemsColor = (command.arguments[1] as? NSObject)?.value(forKey: "toolbar_item_color") as? String
        
        if offerId != nil && !offerId!.isEmpty && toolbasItemsColor != nil && !toolbasItemsColor!.isEmpty {
            Flyy.sharedInstance.openFlyyCustomInviteAndEarnVC(navigationController: self.viewController.navigationController!, offerId: Int(offerId!)!, headerColor: toolbasItemsColor!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected two non-empty string argument."
            )
        }
    }
    
    @objc(setThemeColor:)
    func setThemeColor(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let colorPrimary = (command.arguments[0] as? NSObject)?.value(forKey: "color_primary") as? String
        let colorPrimaryDark = (command.arguments[1] as? NSObject)?.value(forKey: "color_primary_dark") as? String
        
        if colorPrimary != nil && !colorPrimary!.isEmpty && colorPrimaryDark != nil && !colorPrimaryDark!.isEmpty {
            Flyy.sharedInstance.setThemeColor(colorPrimary: colorPrimary!, colorPrimaryDark: colorPrimaryDark!)
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "true"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected two non-empty string argument."
            )
        }
    }
    
    @objc(getShareData:)
    func getShareData(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let offerId = (command.arguments[0] as? NSObject)?.value(forKey: "offer_id") as? String
        
        if offerId != nil && !offerId!.isEmpty {
            Flyy.sharedInstance.getShareData(onComplete: { (success, referralDetails) -> Void in
                if(success) {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_OK,
                        messageAs: referralDetails
                    )
                } else {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_ERROR,
                        messageAs: referralDetails
                    )
                }
            })
            
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(getReferralCount:)
    func getReferralCount(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        Flyy.sharedInstance.getReferralCount(onComplete: {(success, referralCount) -> Void in
            if(success) {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: referralCount
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Failed to fetch data."
                )
            }
        })
    }
    
    @objc(getScratchCardCount:)
    func getScratchCardCount(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        Flyy.sharedInstance.getScratchCardCount(onComplete: {
            (success, scratchCardCountData) -> Void in
            if(success) {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: scratchCardCountData
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: scratchCardCountData
                )
            }
        })
    }
    
    @objc(getPreviousLeaderboardWinners:)
    func getPreviousLeaderboardWinners(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let tag = (command.arguments[0] as? NSObject)?.value(forKey: "tag") as? String
        
        if tag != nil && !tag!.isEmpty {
            Flyy.sharedInstance.getPreviousLeaderboardWinners(tag: tag!, onComplete: {
                (success, message, participantsCount, winnersArray, previousWinnersArray) -> Void in
                if(success) {
                    let data = [participantsCount, winnersArray, previousWinnersArray] as [Any]
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_OK,
                        messageAs: data
                    )
                } else {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_ERROR,
                        messageAs: message
                    )
                }
            })
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(getLeaderboardParticipants:)
    func getLeaderboardParticipants(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let tag = (command.arguments[0] as? NSObject)?.value(forKey: "tag") as? String
        
        if tag != nil && !tag!.isEmpty {
            Flyy.sharedInstance.getLeaderboardParticipants(tag: tag!, onComplete: {
                (success, message, participantsCount, winnersArray, previousWinnersArray) -> Void in
                if(success) {
                    let data = [participantsCount, winnersArray, previousWinnersArray] as [Any]
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_OK,
                        messageAs: data
                    )
                } else {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_ERROR,
                        messageAs: message
                    )
                }
            })
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(getWalletBalance:)
    func getWalletBalance(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        let walletLabel = (command.arguments[0] as? NSObject)?.value(forKey: "wallet_label") as? String
        
        if walletLabel != nil && !walletLabel!.isEmpty {
            Flyy.sharedInstance.getWalletBalance(walletLabel: walletLabel!, onComplete: {
                (success, walletBalanceData) -> Void in
                if(success) {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_OK,
                        messageAs: walletBalanceData
                    )
                } else {
                    pluginResult = CDVPluginResult(
                        status: CDVCommandStatus_ERROR,
                        messageAs: walletBalanceData
                    )
                }
            })
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Expected one non-empty string argument."
            )
        }
    }
    
    @objc(getReferrerDetails:)
    func getReferrerDetails(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        
        Flyy.sharedInstance.getReferrerDetails(onComplete: {
            (success, referrerData) -> Void in
            if(success) {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: referrerData
                )
            } else {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: referrerData
                )
            }
        })
    }
    
    @objc(trackUIEvents:)
    func trackUIEvents(command: CDVInvokedUrlCommand) {
        
    }
}
